cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -125,10 +166,77 @@ public PipelinePlan plan(PipelineSpec spec) {
+   * @param phases the Map of phases created so far
+   * @param phaseConnections the Set of connections between phases added so far
+   * @param outgoingActionConnections the Map that holds set of stages to which
+   *                                  there is an outgoing connection from a Action stage
+   * @param incomingActionConnections the Map that holds set of stages to which
+   *                                  there is a incoming connection to an Action stage
+   * @param subdags subdags created so far from the pipeline stages
+   */
+  private void populateActionPhases(Map<String, StageSpec> specs, Set<String> actionNodes,
+                                    Map<String, PipelinePhase> phases, Set<Connection> phaseConnections,
+                                    Map<String, Set<String>> outgoingActionConnections,
+                                    Map<String, Set<String>> incomingActionConnections, Map<String, Dag> subdags) {
+
+    // Create single stage phases for the Action nodes
+    for (String node : actionNodes) {
+      StageSpec actionStageSpec = specs.get(node);
+      StageInfo actionStageInfo = new StageInfo(node, actionStageSpec.getInputs(), actionStageSpec.getInputSchemas(),
+                                                actionStageSpec.getOutputs(), actionStageSpec.getOutputSchema(),
+                                                actionStageSpec.getErrorDatasetName());
+
+      Map<String, Set<StageInfo>> stages  = new HashMap<>();
+      Set<StageInfo> actionStageInfos = new HashSet<>();
+      actionStageInfos.add(actionStageInfo);
+      stages.put(node, actionStageInfos);
+      phases.put(node, new PipelinePhase(stages, null));
+    }
+
+    // Build phaseConnections for the Action nodes
+    for (Map.Entry<String, Set<String>> connectionFromAction : outgoingActionConnections.entrySet()) {
+
+      // Check if destination is one of the source stages in the pipeline
+      for (Map.Entry<String, Dag> subdagEntry : subdags.entrySet()) {
+        if (Sets.intersection(connectionFromAction.getValue(), subdagEntry.getValue().getSources()).size() > 0) {
+          phaseConnections.add(new Connection(connectionFromAction.getKey(), subdagEntry.getKey()));
+        }
+      }
+
+      // Check if destination is other Action node
+      for (String destination : connectionFromAction.getValue()) {
+        if (actionNodes.contains(destination)) {
+          phaseConnections.add(new Connection(connectionFromAction.getKey(), destination));
+        }
+      }
+    }
+
+    // At this point we have build phaseConnections from Action node to another Action node or phaseConnections
+    // from Action node to another subdags. However it is also possible that sudags connects to the action node.
+    // Build those connections here.
+    for (Map.Entry<String, Set<String>> connectionToAction : incomingActionConnections.entrySet()) {
+      // Check if source is one of the source stages in the pipeline
```

Comment:
seems like it should be // Check if source is one of the sink stages in the pipeline